go/private/actions/compile.bzl

```diff
@@ -53,9 +53,9 @@ def emit_compile(
             fail("compile does not accept deps in bootstrap mode")
         return _bootstrap_compile(go, sources, out_lib, gc_goopts)
 
-    inputs = (sources + [go.package_list] +
-              [archive.data.file for archive in archives] +
-              go.sdk_tools + go.stdlib.files)
+    inputs = sets.union(sources, [go.package_list],
```

Comment:
It is totally possible that my understanding is incorrect, but I believe that there are these benefits of using a depset in such a situation:
1. Memory savings - list is not concatenated when the rule is executed and hence the memory is not allocated to hold it. Instead, depset is just holding references to parts (lists/dpsets).
2. Execution time savings - I assume that constructing a depset is cheap, faster than list concatenation.

Eventually, of course, depset needs to be expanded into a (deduplicated) list and that's when the price is paid - both for deduplication and for memory allocation. However, that price is only paid if that action needs to be executed and never otherwise. And most of the time many actions do not need to run - caching or the output is not required. That is why I believe it is much better to use depsets - memory/cpu cost of analysis (or loading? not sure) phase is reduced and cost is postponed to the execution phase, but only for those actions where it needs to be paid (i.e. where the output is needed).

I may be operating under a wrong assumption here, but I don't see how __constructing__ a depset can be expensive - it would not make any sense, I think. [Performance](https://docs.bazel.build/versions/master/skylark/depsets.html#performance) section does not say it directly, but I read it as if it says that `to_list()` is expensive, not construction of a depset.

Reply:
Sadly, implementation of `depset` is inefficient. It's good to use if you know the asymptotic complexity is strictly better, but compared with something that with the same complexity (list concatenation in thise case), it's almost always worse by a constant factor.

The underlying Java class is [NestedSet](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/collect/nestedset/NestedSet.java;l=43?q=NestedSet&ss=bazel). That one is used directly within Bazel. [SkylarkNestedSet](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/SkylarkNestedSet.java;l=88?q=NestedSet&ss=bazel) is a layer on top of that.

When you create a `NestedSet`, the constructor will de-duplicate the direct elements and the transitive sets using a hash table, so construction is O(len(direct) + len(transitive)) in time and space. Several copies are made, and all the direct elements are hashed at least once, so it's a pretty high constant factor, too.

The crucial thing is that the len(transitive) above is the number of transitive sets passed to the constructor, not the number of elements in those sets. So when you're building a big subgraph of values for linking, it's much more efficient than list concatenation.

Reply:
Thanks for going into details and explaining! I appreciate that very much. I'll undo the change here and in other places.